Replace use of deprecated <ext/hash_map>

<ext/hash_map> has been deprecated.

```
In file included from src/dictionary/PlainDictionary.cpp:34:
In file included from src/dictionary/PlainDictionary.hpp:46:
In file included from /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/ext/hash_map:60:
/usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/backward/backward_warning.h:32:2: warning: This file includes
      at least one deprecated or antiquated header which may be removed without further notice at a future date. Please
      use a non-deprecated interface with equivalent functionality instead. For a listing of replacement headers and
      interfaces, consult the file backward_warning.h. To disable this warning use -Wno-deprecated. [-W#warnings]
#warning \
 ^
```

--- a/libcds/include/libcdsTrees.h
+++ b/libcds/include/libcdsTrees.h
@@ -78,7 +78,7 @@ namespace cds_utils
 	}
 
 	/* writes e[p..p+len-1] = s, len <= W*/
-	inline void bitwrite (register uint *e, register uint p, register uint len, register uint s) {
+	inline void bitwrite (uint *e, uint p, uint len, uint s) {
 		e += p/W;
 		p %= W;
 		if (len == W) {

--- a/libcds/src/static/coders/huff.cpp
+++ b/libcds/src/static/coders/huff.cpp
@@ -139,8 +139,7 @@ namespace cds_static
 		return H;
 	}
 
-	void bitzero (register uint *e, register uint p,
-	register uint len) {
+	void bitzero (uint *e, uint p, uint len) {
 
 		e += p/W; p %= W;
 		if (p+len >= W) {

--- a/libcds/src/static/sequence/SequenceAlphPart.cpp
+++ b/libcds/src/static/sequence/SequenceAlphPart.cpp
@@ -307,7 +307,6 @@ namespace cds_static
 		seqsBuilder->unuse();
 
 		this->cut = cut;
-		this->maxLen = maxLen;
 		this->sigma = sigma;
 	}
 

--- a/libcds/src/static/sequence/SequenceGMRChunk.cpp
+++ b/libcds/src/static/sequence/SequenceGMRChunk.cpp
@@ -66,7 +66,6 @@ namespace cds_static
 		delete [] X_bitmap;
 		this->permutation = pmb->build(pi,chunk_length);
 		assert(permutation!=NULL);
-		this->sigma = sigma;
 		delete [] counter;
 		pmb->unuse();
 		bmb->unuse();
@@ -112,7 +111,6 @@ namespace cds_static
 		delete [] X_bitmap;
 		this->permutation = pmb->build(pi,chunk_length);
 		assert(permutation!=NULL);
-		this->sigma = sigma;
 		delete [] counter;
 		pmb->unuse();
 		bmb->unuse();

--- a/libcds/src/static/sequence/wt_coder_huff.cpp
+++ b/libcds/src/static/sequence/wt_coder_huff.cpp
@@ -76,7 +76,7 @@ namespace cds_static
         uint *buffer = new uint[maxBuffer];
         hc->encode(symbol, buffer, (size_t)0);
         bool ret = bitget(buffer,l);
-        delete buffer;
+        delete[] buffer;
         return ret;
     }
 
@@ -89,7 +89,7 @@ namespace cds_static
         uint *buffer = new uint[maxBuffer];
         uint s_len = (uint)hc->encode(symbol, buffer, (size_t)0);
         bool ret =  l==s_len;
-        delete buffer;
+        delete[] buffer;
         return ret;
     }
 

--- a/libcds/src/static/suffixtree/RMQ_succinct.cpp
+++ b/libcds/src/static/suffixtree/RMQ_succinct.cpp
@@ -104,7 +104,7 @@ namespace cds_static
 	unsigned int RMQ_succinct::log2fast(unsigned int v) {
 		unsigned int c = 0;		 // c will be lg(v)
 								 // temporaries
-		register unsigned int t, tt;
+		unsigned int t, tt;
 
 		if ((tt = v >> 16))
 			c = (t = v >> 24) ? 24 + LogTable256[t] : 16 + LogTable256[tt & 0xFF];

--- a/libcds/src/static/suffixtree/RMQ_succinct_lcp.cpp
+++ b/libcds/src/static/suffixtree/RMQ_succinct_lcp.cpp
@@ -103,7 +103,7 @@ namespace cds_static
 
 	uint RMQ_succinct_lcp::log2fast(uint v) {
 		uint c = 0;				 // c will be lg(v)
-		register uint t, tt;	 // temporaries
+	        uint t, tt;	 // temporaries
 
 		if ((tt = v >> 16))
 			c = (t = v >> 24) ? 24 + LogTable256[t] : 16 + LogTable256[tt & 0xFF];

--- a/libcds/src/static/suffixtree/factorization.cpp
+++ b/libcds/src/static/suffixtree/factorization.cpp
@@ -58,7 +58,7 @@ namespace cds_static
 		uint levelSizeAux[7]={0,0,0,0,0,0,0};
 		uint cont[7]={0,0,0,0,0,0,0};
 		listLength = l_Length;
-		register uint i;
+		uint i;
 		int j, k;
 		uint value, newvalue;
 		uint bits_BS_len = 0;
@@ -139,7 +139,7 @@ namespace cds_static
 
 	uint factorization::access(uint param) {
 		uint mult=0;
-		register uint j;
+		uint j;
 		uint partialSum=0;
 		uint ini=param-1;
 		uint * bsData = ((BitSequenceRG *)bS)->data;
@@ -173,7 +173,7 @@ namespace cds_static
 
 	uint factorization::access_seq(uint param, size_t *next_pos, bool dir) {
 		uint mult=0;
-		register uint j;
+	        uint j;
 		uint partialSum=0;
 		uint ini=param-1;
 		uint * bsData = ((BitSequenceRG *)bS)->data;

--- a/libcds/src/static/suffixtree/factorization_var.cpp
+++ b/libcds/src/static/suffixtree/factorization_var.cpp
@@ -41,7 +41,7 @@ namespace cds_static
 		uint *cont;
 		uint *contB;
 		listLength = l_Length;
-		register uint i;
+	        uint i;
 		int j, k;
 		uint value, newvalue;
 		uint bits_BS_len = 0;
@@ -171,7 +171,7 @@ namespace cds_static
 
 	uint factorization_var::access(uint param) {
 		uint mult=0;
-		register uint j;
+		uint j;
 		uint partialSum=0;
 		uint ini = param-1;
         unsigned char readByte;
@@ -200,7 +200,7 @@ namespace cds_static
 
 	uint factorization_var::access_seq(uint param, size_t *next_pos, bool dir) {
 		uint mult=0;
-		register uint j;
+		uint j;
 		uint partialSum=0;
 		uint ini = param-1;
         unsigned char readByte;

--- a/libhdt/src/dictionary/PlainDictionary.hpp
+++ b/libhdt/src/dictionary/PlainDictionary.hpp
@@ -41,12 +41,7 @@
 #include <algorithm>
 #include <fstream>
 #include <iostream>
-
-#ifndef WIN32
-#include <ext/hash_map>
-#else
 #include <unordered_map>
-#endif
 
 //#define GOOGLE_HASH
 
@@ -88,13 +83,8 @@ typedef std::pair<const char*, DictionaryEntry *> DictEntryPair;
 typedef sparse_hash_map<const char *, DictionaryEntry *, hash<const char *>, str_cmp> DictEntryHash;
 #else
 
-#ifdef WIN32
-/*typedef std::hash_map<const char *, DictionaryEntry *, hash<const char *>, str_cmp> DictEntryHash;*/
 typedef unordered_map<const char *, DictionaryEntry *, hash<const char *>, str_cmp> DictEntryHash;
-#else
-typedef std::hash_map<const char *, DictionaryEntry *, __gnu_cxx::hash<const char *>, str_cmp> DictEntryHash;
 
-#endif
 #endif
 
 typedef DictEntryHash::const_iterator DictEntryIt;

--- a/libhdt/src/dictionary/PlainDictionary.hpp
+++ b/libhdt/src/dictionary/PlainDictionary.hpp
@@ -43,23 +43,6 @@
 #include <iostream>
 #include <unordered_map>
 
-//#define GOOGLE_HASH
-
-#ifdef GOOGLE_HASH
-#include <sparsehash/sparse_hash_map>
-
-using google::sparse_hash_map;      // namespace where class lives by default
-using __gnu_cxx::hash;  // or __gnu_cxx::hash, or maybe tr1::hash, depending on your OS
-
-#else
-
-#ifndef WIN32
-namespace std { using namespace __gnu_cxx; }
-#endif
-
-#endif
-
-
 namespace hdt {
 
 struct DictionaryEntry {
@@ -79,16 +62,9 @@ struct str_cmp {
 
 typedef std::pair<const char*, DictionaryEntry *> DictEntryPair;
 
-#ifdef GOOGLE_HASH
-typedef sparse_hash_map<const char *, DictionaryEntry *, hash<const char *>, str_cmp> DictEntryHash;
-#else
-
-typedef unordered_map<const char *, DictionaryEntry *, hash<const char *>, str_cmp> DictEntryHash;
-
-#endif
+typedef std::unordered_map<const char *, DictionaryEntry *, hash<std::string>, str_cmp> DictEntryHash;
 
 typedef DictEntryHash::const_iterator DictEntryIt;
-
 
 class PlainDictionary : public ModifiableDictionary {
 private:

--- a/libhdt/src/hdt/BasicModifiableHDT.cpp
+++ b/libhdt/src/hdt/BasicModifiableHDT.cpp
@@ -152,8 +152,6 @@ void BasicModifiableHDT::saveToHDT(std::ostream & output, ProgressListener *list
 	st.reset();
 	triples->save(output, controlInformation);
 	cerr << "Triples saved in " << st << endl;
-
-	this->fileName = fileName;
 }
 
 void BasicModifiableHDT::insert(TripleString & triple)

--- a/libhdt/src/huffman/huff.cpp
+++ b/libhdt/src/huffman/huff.cpp
@@ -162,8 +162,7 @@ namespace URICompressed{
         return H;
     }
 
-    void bitzero (register uint *e, register uint64_t p,
-    register uint len) {
+    void bitzero (uint *e, uint64_t p, uint len) {
 
         e += p/W; p %= W;
         if (p+len >= W) {

--- a/libhdt/src/libdcs/CSD_PFC.h
+++ b/libhdt/src/libdcs/CSD_PFC.h
@@ -188,7 +188,6 @@ class PFCSuggestionIterator: public hdt::IteratorUCharString {
 private:
 	CSD_PFC *pfc;
 	size_t max;
-	size_t count;
 	const char *prefix;
 	bool hasnext;
 	string tmpStr;
@@ -325,7 +324,6 @@ class PFCSuggestionIDIterator: public hdt::IteratorUInt {
 private:
 	CSD_PFC *pfc;
 	size_t max;
-	size_t count;
 	const char *prefix;
 	bool hasnext;
 	string tmpStr;

--- a/libhdt/src/libdcs/CSD_PFC.h
+++ b/libhdt/src/libdcs/CSD_PFC.h
@@ -201,7 +201,7 @@ private:
 public:
 
 	PFCSuggestionIterator(CSD_PFC *pfc, const char *prefix) :
-			pfc(pfc), count(1), prefix(prefix) {
+			pfc(pfc), prefix(prefix) {
 		max = pfc->getLength();
 		prefixlen = strlen(prefix);
 		hasnext = false;
@@ -337,7 +337,7 @@ private:
 public:
 
 	PFCSuggestionIDIterator(CSD_PFC *pfc, const char *prefix) :
-			pfc(pfc), count(1), prefix(prefix) {
+			pfc(pfc), prefix(prefix) {
 		max = pfc->getLength();
 		prefixlen = strlen(prefix);
 		hasnext = false;

--- a/libhdt/src/sequence/HuffmanSequence.cpp
+++ b/libhdt/src/sequence/HuffmanSequence.cpp
@@ -87,7 +87,7 @@ void HuffmanSequence::load(std::istream & input)
     for(size_t i=0;i<numElements;i++) {
 		pos = huffman->decode(&vectorPlain[i], encV, pos);
 	}
-    delete encV;
+    delete[] encV;
 }
 
 size_t HuffmanSequence::load(const unsigned char *ptr, const unsigned char *ptrMax, ProgressListener *listener)

